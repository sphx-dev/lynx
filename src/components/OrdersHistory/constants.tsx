//TODO: change <span> to components when data & designs will be clearer
export const mockColumns = [
  {
    accessorKey: "market",
    header: "Market",
    cell: (props: any) => <span>{props.getValue()}</span>,
  },
  {
    accessorKey: "size",
    header: "Size",
    cell: (props: any) => <span>{props.getValue()}</span>,
  },
  {
    accessorKey: "slide",
    header: "Slide",
    cell: (props: any) => <span>{props.getValue()}</span>,
  },
  {
    accessorKey: "avgEntry",
    header: "Avg Entry",
    cell: (props: any) => <span>{props.getValue()}</span>,
  },
  {
    accessorKey: "liqPrice",
    header: "Liq Price",
    cell: (props: any) => <span>{props.getValue()}</span>,
  },
  {
    accessorKey: "marketMargin",
    header: "Market Margin",
    cell: (props: any) => <span>{props.getValue()}</span>,
  },
  {
    accessorKey: "u&pl",
    header: "U&PL",
    cell: (props: any) => <span>{props.getValue()}</span>,
  },
]

export const mockData = [
  {
    market: "BTC/USDT",
    size: "0.001",
    slide: "0.001",
    avgEntry: "0.001",
    liqPrice: "0.001",
    marketMargin: "0.001",
    "u&pl": "0.001",
  },
]
