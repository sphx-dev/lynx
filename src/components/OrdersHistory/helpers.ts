import { Side } from "../../types/order";
import Colors from "../../theme/colors";

export const getSideColor = (side: Side) =>
  side === Side.Buy ? Colors.common.positive1 : Colors.common.negative3;
