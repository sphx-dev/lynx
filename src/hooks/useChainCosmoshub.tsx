import {
  useAccount,
  useConnect,
  useDisconnect,
  // useStargateClient,
  WalletType,
} from "graz";
import { Keplr } from "@keplr-wallet/types";
import { getChain } from "../config";

// export const useChainCosmoshub = () => useChain("cosmoshub");
export const useChainCosmoshub = () => {
  const { connectAsync, status: connectionStatus, error } = useConnect();
  const {
    data: account,
    isConnected,
    status: accountStatus,
    reconnect,
  } = useAccount();
  const { disconnect } = useDisconnect();

  // TODO: use client to get sequence if needed for order creation
  // const { data: client } = useStargateClient();

  // (async () => {
  //   console.log("ACCOUNT", account);
  //   console.log("CLIENT", client);
  //   console.log("CLIENT seq", client?.getSequence);
  //   console.log(
  //     "CLIENT seq addr",
  //     await client?.getSequence(account?.bech32Address)
  //   );
  // })();

  if (error) {
    console.log("useConnect error", error);
  }

  return {
    addChain: async () => {
      try {
        const keplr: Keplr = window.keplr;
        await keplr.experimentalSuggestChain(getChain());
      } catch (e) {
        console.log("addChain ERR -------->", e);
      }
    },
    connect: async () => {
      try {
        await connectAsync({
          chainId: getChain().chainId,
          walletType: WalletType.KEPLR,
        });
      } catch (e) {
        console.log("-------->", e);
      }
    },
    reconnect,
    disconnect,
    connectionStatus,
    accountStatus,
    account,
    address: account?.bech32Address,
    isConnected,
  };
};
