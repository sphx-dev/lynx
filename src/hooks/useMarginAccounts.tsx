import {
  createMarginAccount,
  getAccountsByOwner,
} from "../utils/queryMarginAccounts";
import { MarginAccountInfo } from "proto-codecs/codegen/sphx/marginacc/margin_account";
import { create } from "zustand";
import { useCallback, useEffect } from "react";
import { errorAlert, successAlert } from "@/utils/alerts";
import { useMutation, useQuery, useQueryClient } from "react-query";
import { useTranslation } from "react-i18next";

type MarginAccountState = {
  marginAccounts: MarginAccountInfo[];
  selectedIndex: number;

  setMarginAccounts: (marginAccounts: MarginAccountInfo[]) => void;
  setSelectedIndex: (selectedIndex: number) => void;

  getSelectedAccount: () => MarginAccountInfo | undefined;
  getSelectedAddress: () => string | undefined;
  getSelectedIndex: () => number;
};

export const useMarginAccount = (owner: string | null | undefined) => {
  const { t } = useTranslation();
  const store = useMarginAccountStore();
  const { setMarginAccounts } = store;

  const queryClient = useQueryClient();

  // Fetch margin accounts
  const { data, isLoading, isError, isIdle, isSuccess, status } = useQuery(
    ["marginAccounts", owner],
    () => {
      return getAccountsByOwner(owner!).then(response => {
        return response.marginAccounts;
      });
    },
    {
      enabled: !!owner,
      staleTime: 1000 * 60 * 60,
    }
  );
  useEffect(() => {
    if (data) {
      setMarginAccounts(data);
    }
  }, [data, setMarginAccounts]);

  // Mutation for creating margin account
  const mutation = useMutation(createMarginAccount);
  const createAccount = useCallback(() => {
    const ids = store.marginAccounts.map(ma => ma?.id?.number ?? 0);
    const nextMarginAccountNumber = Math.max(...ids, 0) + 1;
    return mutation.mutateAsync(
      {
        address: owner!,
        number: nextMarginAccountNumber,
        memo: t("createMarginAccount"),
      },
      {
        onSettled: () => {
          queryClient.invalidateQueries(["marginAccounts", owner]);
        },
        onSuccess: response => {
          if (response.code === 0) {
            successAlert("Margin account created successfully");
          } else {
            errorAlert(getTextByCode(response.code));
          }
        },
        onError: error => {
          errorAlert(getTextFromError((error as Error).message));
        },
      }
    );
  }, [t, owner, mutation, store.marginAccounts, queryClient]);

  return {
    ...store,
    status,
    isIdle,
    isLoading,
    isError,
    isSuccess,
    selectedAccount: store.getSelectedAccount(),
    selectedAddress: store.getSelectedAddress(),
    selectedIndex: store.getSelectedIndex(),
    createAccount,
  };
};

export const useMarginAccountStore = create<MarginAccountState>((set, get) => ({
  marginAccounts: [],
  selectedIndex: 0,

  // Setters
  setMarginAccounts: (marginAccounts: MarginAccountInfo[]) => {
    set({
      marginAccounts,
    });
  },
  setSelectedIndex: (selectedIndex: number) => {
    set({
      selectedIndex,
    });
  },

  // Getters
  getSelectedAccount: () => {
    return get().marginAccounts[get().selectedIndex];
  },
  getSelectedAddress: () => {
    return get().marginAccounts[get().selectedIndex]?.address;
  },
  getSelectedIndex: () => {
    return get().selectedIndex;
  },
}));

//TODO: move this to translations
const getTextByCode = (code: number) => {
  switch (code) {
    case 9:
      return "Margin account already exists";
    default:
      return "Error creating margin account. Please try again";
  }
};

const getTextFromError = (message: string) => {
  if (
    message.includes("Send some tokens there before trying to query sequence")
  ) {
    return "Seems like youâ€™re using new wallet. You must have at least 0.01 USDC for transaction fee";
  }

  return "Something went wrong. Please try again";
};
