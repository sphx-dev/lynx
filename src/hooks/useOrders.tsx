import { getOrderByOrderId, getOrdersByAddress } from "../utils/queryOrders";

import {
  cancelOrderInChain,
  placeLimitOrderInChain,
  PlaceLimitOrderInChainParams,
  placeMarketOrderInChain,
  PlaceMarketOrderInChainParams,
} from "../utils/placeOrder";

import { useMutation, useQueries, useQuery, useQueryClient } from "react-query";
import { OrderStatus } from "proto-codecs/codegen/sphx/order/validated_order";
import { OrderId } from "proto-codecs/codegen/sphx/order/order";

const PAGE_SIZE = 10n;

export const useOrders = (
  address: string | undefined,
  page: number,
  status?: OrderStatus
) => {
  const {
    isLoading,
    error,
    data: { orders, totalOrders } = { orders: [], totalOrders: 0 },
  } = useQuery(
    ["orders", address, page, status],
    () => {
      return getOrdersByAddress(
        address!,
        { offset: BigInt(page) * PAGE_SIZE, limit: PAGE_SIZE },
        status
      ).then(response => {
        // console.log("useOrders response", response);
        return {
          orders: response?.orders,
          totalOrders: Number(response?.pagination?.total) || 0,
        };
      });
    },
    {
      enabled: !!address,
      staleTime: 1000 * 60,
    }
  );

  return {
    isLoading,
    error,
    orders,
    totalOrders,
    pageSize: Number(PAGE_SIZE),
  };
};

export const useOrdersById = (orderIds: OrderId[]) => {
  const results = useQueries(
    orderIds.map(orderId => ({
      queryKey: [
        "order",
        orderId.marginAccountAddress,
        orderId.number.toString(),
      ],
      queryFn: async () => {
        const respose = await getOrderByOrderId(orderId);
        console.log("useOrdersById response", respose);
        return respose.order;
      },
      enabled:
        !!orderId && !!orderId?.marginAccountAddress && !!orderId?.number,
      staleTime: 1000 * 60,
    }))
  );

  return results;
};

export const useCancelOrder = () => {
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: cancelOrderInChain,
    onSettled: async () => {
      queryClient.invalidateQueries({ queryKey: ["orders"] });
    },
  });

  const cancelOrder = (params: {
    address: string;
    orderId: OrderId;
    memo: string;
  }) => {
    console.log("cancelOrder PARAMS", params);

    return mutation.mutateAsync(params, {
      onSettled: () => {
        queryClient.invalidateQueries("orders");
      },
    });
  };
  return { cancelOrder };
};

export const useCreateOrder = () => {
  const queryClient = useQueryClient();
  const placeMarketMutation = useMutation({
    mutationFn: placeMarketOrderInChain,
  });
  const placeMarketOrder = async (params: PlaceMarketOrderInChainParams) => {
    return placeMarketMutation.mutateAsync(params, {
      onSettled: () => {
        queryClient.invalidateQueries("orders");
      },
    });
  };

  const placeLimitMutation = useMutation({
    mutationFn: placeLimitOrderInChain,
  });
  const placeLimitOrder = async (params: PlaceLimitOrderInChainParams) => {
    return placeLimitMutation.mutateAsync(params, {
      onSettled: () => {
        queryClient.invalidateQueries("orders");
      },
    });
  };

  return {
    placeMarketOrder,
    marketStatus: placeMarketMutation,
    placeLimitOrder,
    limitStatus: placeLimitMutation,
  };
};
