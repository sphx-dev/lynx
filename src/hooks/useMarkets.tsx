import { create } from "zustand";
import { Market } from "../../proto-codecs/codegen/sphx/order/market";
import { getAllMarkets } from "../utils/queryMarkets";
import { useEffect, useMemo } from "react";
import { useQuery } from "react-query";
import { marketPriority, marketsIcons } from "@/constants/marketsIcons";

export const useMarkets = () => {
  const { markets, selectedMarket, setMarkets, setMarketId } =
    useMarketsStore();

  const { data } = useQuery("markets", getAllMarkets, { staleTime: Infinity });
  useEffect(() => {
    let m: Market[] = [];
    if (data?.markets) {
      m = data.markets.toSorted((a, b) => {
        return (
          (marketPriority[b?.ticker] ?? 0) - (marketPriority[a?.ticker] ?? 0)
        );
      });
    }
    setMarkets(m);
  }, [data, setMarkets]);

  const { icon } = useMemo(() => {
    if (selectedMarket?.ticker)
      if (selectedMarket?.ticker in marketsIcons) {
        return marketsIcons[selectedMarket?.ticker];
      }
    return marketsIcons["default"];
  }, [selectedMarket]);

  window.marketId = selectedMarket?.id || 0n;

  let selectedMarketMeta: any = {};
  if (selectedMarket?.ticker) {
    selectedMarketMeta = MARKETS_META[selectedMarket?.ticker] || {};
  }

  return {
    markets,
    selectedMarket,
    selectedMarketId: selectedMarket?.id,
    symbol: selectedMarket?.ticker,
    icon,
    setMarkets,
    setMarketId,
    minimumVolume: selectedMarketMeta?.minimumVolume || 0,
    pricePerContract: selectedMarketMeta?.pricePerContract || 1,
  };
};

type MarketsStore = {
  markets: Market[];
  selectedMarketId: bigint;
  selectedMarket: Market | null;
  setMarkets: (markets: Market[]) => void;
  setMarketId: (id: bigint) => void;
};

const useMarketsStore = create<MarketsStore>((set, get) => ({
  markets: [],
  selectedMarketId: getLocalStorageSelectedMarket(),
  selectedMarket: null,
  setMarkets: (markets: Market[]) => {
    if (equal(get().markets, markets)) return;

    const isSelectedMarketInMarkets = markets.some(
      m => m.id === get().selectedMarketId
    );
    const selectedId = isSelectedMarketInMarkets
      ? get().selectedMarketId
      : markets[0]?.id || 0n;

    setLocalStorageSelectedMarket(selectedId);
    set({
      markets,
      selectedMarketId: selectedId,
      selectedMarket: markets.find(m => m.id === selectedId),
    });
  },
  setMarketId: id => {
    const selectedMarket = get().markets.find(market => market.id === id);
    if (selectedMarket) {
      set({ selectedMarketId: id, selectedMarket });
      setLocalStorageSelectedMarket(id);
    }
  },
}));

const equal = (a: Market[], b: Market[]) => {
  if (a.length === 0 && b.length === 0) return true;
  if (a.length !== b.length) return false;
  return a.every(
    (m, i) =>
      m.id === b[i]?.id &&
      m.ticker === b[i]?.ticker &&
      m.status === b[i]?.status &&
      m.baseAsset === b[i]?.baseAsset &&
      m.quoteAsset === b[i]?.quoteAsset
  );
};

const MARKETS_META: Record<string, any> = {
  "BTCUSDC.P": {
    minimumVolume: 0,
    pricePerContract: 1,
  },
  "ETHUSDC.P": {
    minimumVolume: 0,
    pricePerContract: 1,
  },
  "WTXUSDC.P": {
    minimumVolume: 0.1,
    pricePerContract: 10,
  },
};

function getLocalStorageSelectedMarket() {
  let strMarket = localStorage.getItem("selectedMarket") ?? 0;
  try {
    return BigInt(strMarket);
  } catch (e) {
    return BigInt(0);
  }
}

function setLocalStorageSelectedMarket(id: bigint) {
  localStorage.setItem("selectedMarket", id.toString());
}
