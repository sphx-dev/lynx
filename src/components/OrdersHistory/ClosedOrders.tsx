import React from "react";
import { dateToDisplay } from "../../utils/date";
import Text from "../Text";
import Table from "./Table";
import { useAppSelector } from "../../hooks";
import { account } from "../../state/accountSlice";
import { getSideColor } from "./helpers";
import PlaceHolder from "./PlaceHolder";
import { formatNumber } from "../../utils/format";

const ClosedOrders = () => {
  const { closedOrders } = useAppSelector(account);
  const columns = [
    {
      accessorKey: "timestamp",
      header: "Date",
      cell: (props: any) => (
        <Text color="tertiary">{dateToDisplay(props.getValue())}</Text>
      ),
    },
    {
      accessorKey: "quantity",
      header: "Size",
      cell: (props: any) => <Text color="tertiary">{props.getValue()}</Text>,
    },
    {
      accessorKey: "price",
      header: "Price",
      cell: (props: any) => (
        <Text color="tertiary">
          {formatNumber({ value: +props.getValue(), fixed: 2 })}
        </Text>
      ),
    },
    {
      accessorKey: "side",
      header: "Side",
      cell: (props: any) => (
        <Text color={getSideColor(props.getValue())}>{props.getValue()}</Text>
      ),
    },
  ];
  if (!closedOrders.length) {
    return <PlaceHolder>No Orders yet</PlaceHolder>;
  }
  return <Table columns={columns} data={closedOrders} />;
};

export default ClosedOrders;
