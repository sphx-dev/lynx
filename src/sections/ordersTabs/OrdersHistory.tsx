import { useMemo, useState } from "react";
import { Text, Table } from "../../components";
import { getOrderStatusText, getOrderTypeText, getSideColor } from "./helpers";
import PlaceHolder from "./PlaceHolder";
import { formatDollars } from "../../utils/format";
import { useTranslation } from "react-i18next";
import { useChainCosmoshub } from "@/hooks/useChainCosmoshub";
import { useMarginAccount } from "@/hooks/useMarginAccounts";
import { useOrders } from "@/hooks/useOrders";
import {
  OrderStatus,
  ValidatedOrder,
} from "proto-codecs/codegen/sphx/order/validated_order";
import { OrderSide } from "proto-codecs/codegen/sphx/order/order";
import { Side } from "@/types/order";
import { Pagination } from "@/components/Pagination";
import { useMarkets } from "@/hooks/useMarkets";
import dayjs from "dayjs";
import { PRECISION } from "@/constants";
import { ReloadButton } from "./components/ReloadButton";
import { LoaderBar } from "@/components/LoaderBar";

const OrdersHistory = () => {
  const { address } = useChainCosmoshub();
  const { selectedAddress } = useMarginAccount(address);

  const columns = useOrdersHistoryColumns();

  const [page, setPage] = useState<number>(0);
  const {
    orders: originalOrders,
    totalOrders,
    pageSize,
    refetch,
    isFetching,
  } = useOrders(selectedAddress, page, [
    OrderStatus.ORDER_STATUS_CANCELED,
    OrderStatus.ORDER_STATUS_FILLED,
    OrderStatus.ORDER_STATUS_EXPIRED,
    OrderStatus.ORDER_STATUS_PARTIALLY_FILLED,
  ]);
  const orders = originalOrders.toReversed();

  if (totalOrders === 0) {
    return (
      <>
        <LoaderBar style={{ visibility: isFetching ? "visible" : "hidden" }} />
        <ReloadButton onClick={() => refetch()} />
        <PlaceHolder>No Orders yet</PlaceHolder>
      </>
    );
  }

  return (
    <>
      <LoaderBar style={{ visibility: isFetching ? "visible" : "hidden" }} />
      <ReloadButton onClick={() => refetch()} />
      <Table columns={columns} data={orders} />
      <Pagination
        page={page}
        setPage={setPage}
        totalItems={totalOrders}
        pageSize={pageSize}
      />
    </>
  );
};

export default OrdersHistory;

const useOrdersHistoryColumns = () => {
  const { t } = useTranslation();
  const { markets } = useMarkets();
  const marketMap = useMemo(() => {
    const map = new Map();
    markets.forEach(m => {
      map.set(m.id, t(m.baseAsset) + "/" + t(m.quoteAsset));
    });
    return map;
  }, [markets, t]);

  return [
    {
      accessorKey: "marketId",
      header: "Market",
      cell: (props: any) => (
        <Text color="tertiary">{marketMap.get(props.getValue())}</Text>
      ),
    },
    {
      flex: 1,
      accessorKey: "timestamp",
      header: "Date",
      cell: (props: any) => (
        <Text color="tertiary">
          {dayjs(Date.now()).diff(
            dayjs(Number(props.getValue()) * 1000),
            "day"
          ) > 1
            ? dayjs(Number(props.getValue()) * 1000).format("YYYY-MM-DD")
            : dayjs(Number(props.getValue()) * 1000).format("HH:mm:ss")}
        </Text>
      ),
    },
    {
      accessorKey: "orderType",
      header: "Type",
      cell: (props: any) => (
        <Text color="tertiary">{getOrderTypeText(props.getValue(), t)}</Text>
      ),
    },
    {
      // accessorKey: "quantity",
      accessorFn: (order: ValidatedOrder) => {
        return (
          order?.fills?.reduce((acc, fill) => {
            return acc + Number(fill.quantity);
          }, 0) || 0
        );
        // return order.quantity;
      },
      header: "Exec. Size",
      cell: (props: any) => (
        <Text color="tertiary">{Number(props.getValue()) / PRECISION}</Text>
      ),
    },
    {
      // accessorKey: "price",
      accessorFn: (order: ValidatedOrder) => {
        let totalQuantity = 0n;

        let accPrice =
          order?.fills?.reduce((acc, fill) => {
            totalQuantity += BigInt(fill.quantity);
            return acc + BigInt(fill.quantity) * BigInt(fill.price);
          }, 0n) || 0n;

        return Number((accPrice / totalQuantity).toString()) / PRECISION;

        // if (order.orderType === OrderType.ORDER_TYPE_MARKET) {
        //   return t("marketPrice");
        // }
        // return formatPrice(Number(order.price) / PRECISION, 2);
      },
      header: "Exec. Price",
      cell: (props: any) => (
        <Text color="tertiary">
          {/* {props.getValue()}/{formatPrice(Number(props.getValue()), 2)}/ */}
          {formatDollars(Number(props.getValue()))}
        </Text>
      ),
    },
    {
      accessorKey: "leverage",
      header: "Lev.",
      cell: (props: any) => (
        <Text color="tertiary">x{Number(props.getValue())}</Text>
      ),
    },
    {
      accessorKey: "side",
      header: "Side",
      cell: (props: any) => (
        <Text
          color={getSideColor(
            props.getValue() === OrderSide.ORDER_SIDE_BUY ? Side.Buy : Side.Sell
          )}
        >
          {props.getValue() === OrderSide.ORDER_SIDE_BUY ? "buy" : "sell"}
        </Text>
      ),
    },
    {
      accessorKey: "status",
      header: "Status",
      cell: (props: any) => (
        <Text>{getOrderStatusText(props.getValue(), t)}</Text>
      ),
    },
  ];
};
