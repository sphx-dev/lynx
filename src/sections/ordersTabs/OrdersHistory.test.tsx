import { render, screen } from "@testing-library/react";
import { describe, test, expect, vi, Mock } from "vitest";
import OrdersHistory from "./OrdersHistory";
import { useChainCosmoshub } from "@/hooks/useChainCosmoshub";
import { useMarginAccount } from "@/hooks/useMarginAccounts";
import { useOrders } from "@/hooks/useOrders";
import { useMarkets } from "@/hooks/useMarkets";
import dayjs from "dayjs";
import { AppWrapper } from "@/testing/AppWrapper";

// Mock hooks
vi.mock("@/hooks/useChainCosmoshub", () => ({
  useChainCosmoshub: vi.fn(),
}));
vi.mock("@/hooks/useMarginAccounts", () => ({
  useMarginAccount: vi.fn(),
}));
vi.mock("@/hooks/useOrders", () => ({
  useOrders: vi.fn(),
}));
vi.mock("@/hooks/useMarkets", () => ({
  useMarkets: vi.fn(),
}));

describe("ClosedOrders", () => {
  beforeEach(() => {
    // Mock date
    vi.useFakeTimers();
    let date = new Date("2024-10-01");
    // this line sets the timezone to UTC
    date.setMinutes(date.getMinutes() - date.getTimezoneOffset());
    vi.setSystemTime(date);

    // Mock hooks
    (useChainCosmoshub as Mock).mockReturnValue({ address: "test-address" });
    (useMarginAccount as Mock).mockReturnValue({
      selectedAddress: "test-selected-address",
    });
    (useOrders as Mock).mockReturnValue({
      orders: [
        {
          marketId: "1",
          timestamp: dayjs("2024-10-01").subtract(2, "days").unix(),
          quantity: 1810000,
          price: 5000000,
          leverage: 2,
          side: 1,
          status: 4,
          orderType: 0,
        },
        {
          marketId: "1",
          timestamp: dayjs("2024-10-01").subtract(3, "days").unix(),
          quantity: 1100000,
          price: 4912000,
          leverage: 2,
          side: 1,
          status: 3,
          orderType: 1,
        },
        {
          marketId: "1",
          timestamp: dayjs("2024-10-01").subtract(3, "days").unix(),
          quantity: 1100000,
          price: 4112000,
          leverage: 2,
          side: 1,
          status: 3,
          orderType: 0,
        },
      ],
      totalOrders: 2,
      pageSize: 10,
    });
    (useMarkets as Mock).mockReturnValue({
      markets: [{ id: "1", baseAsset: "BTC", quoteAsset: "USD" }],
    });
  });

  afterEach(() => {
    vi.useRealTimers();
    vi.clearAllMocks();
  });

  test("renders without crashing", () => {
    const { container } = render(<OrdersHistory />, { wrapper: AppWrapper });

    expect(screen.getAllByText("Market").length).toBe(2);
    expect(screen.getAllByText("Limit").length).toBe(2);
    expect(container.querySelector("table")?.outerHTML).toMatchSnapshot();
  });

  test("displays orders correctly", () => {
    const { container } = render(<OrdersHistory />, { wrapper: AppWrapper });

    const rows = container.querySelectorAll("tr");

    expect(rows[1].querySelectorAll("td")[0].textContent).toBe("BTC/USD");
    expect(rows[1].querySelectorAll("td")[1].textContent).toBe("2024-09-29");
    expect(rows[1].querySelectorAll("td")[2].textContent).toBe("Limit");
    expect(rows[1].querySelectorAll("td")[3].textContent).toBe("1.81");
    expect(rows[1].querySelectorAll("td")[4].textContent).toBe("5.00");
    expect(rows[1].querySelectorAll("td")[5].textContent).toBe("x2");
    expect(rows[1].querySelectorAll("td")[6].textContent).toBe("buy");
    expect(rows[1].querySelectorAll("td")[7].textContent).toBe("Canceled");

    expect(rows[2].querySelectorAll("td")[0].textContent).toBe("BTC/USD");
    expect(rows[2].querySelectorAll("td")[1].textContent).toBe("2024-09-28");
    expect(rows[2].querySelectorAll("td")[2].textContent).toBe("Market");
    expect(rows[2].querySelectorAll("td")[3].textContent).toBe("1.1");
    expect(rows[2].querySelectorAll("td")[4].textContent).toBe("Market Price");
    expect(rows[2].querySelectorAll("td")[5].textContent).toBe("x2");
    expect(rows[2].querySelectorAll("td")[6].textContent).toBe("buy");
    expect(rows[2].querySelectorAll("td")[7].textContent).toBe("Filled");

    expect(rows[3].querySelectorAll("td")[0].textContent).toBe("BTC/USD");
    expect(rows[3].querySelectorAll("td")[1].textContent).toBe("2024-09-28");
    expect(rows[3].querySelectorAll("td")[2].textContent).toBe("Limit");
    expect(rows[3].querySelectorAll("td")[3].textContent).toBe("1.1");
    expect(rows[3].querySelectorAll("td")[4].textContent).toBe("4.11");
    expect(rows[3].querySelectorAll("td")[5].textContent).toBe("x2");
    expect(rows[3].querySelectorAll("td")[6].textContent).toBe("buy");
    expect(rows[3].querySelectorAll("td")[7].textContent).toBe("Filled");
  });

  test("displays placeholder when no orders", () => {
    (useOrders as Mock).mockReturnValue({
      orders: [],
      totalOrders: 0,
      pageSize: 10,
    });
    render(<OrdersHistory />, { wrapper: AppWrapper });
    expect(screen.getByText("No Orders yet")).toBeInTheDocument();
  });
});
