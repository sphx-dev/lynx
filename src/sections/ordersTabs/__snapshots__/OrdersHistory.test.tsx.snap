// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ClosedOrders > renders without crashing 1`] = `"<table class="sc-dstKZu fQDZyI"><thead><tr><th><div class="sc-dpBQxM hlYqZS">Market</div></th><th><div class="sc-dpBQxM hlYqZS">Date</div></th><th><div class="sc-dpBQxM hlYqZS">Type</div></th><th><div class="sc-dpBQxM hlYqZS">Size</div></th><th><div class="sc-dpBQxM hlYqZS">Price</div></th><th><div class="sc-dpBQxM hlYqZS">Lev.</div></th><th><div class="sc-dpBQxM hlYqZS">Side</div></th><th><div class="sc-dpBQxM hlYqZS">Status</div></th></tr></thead><tbody><tr class="sc-keTIit hZSLjO"><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">BTC/USD</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">2024-09-29</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">Limit</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">1.81</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">5.00</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">x2</p></td><td><p align="left" color="#4C972E" class="sc-blHHSb cUlyYr">buy</p></td><td><p align="left" class="sc-blHHSb bXnAcp">Canceled</p></td></tr><tr class="sc-keTIit hZSLjO"><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">BTC/USD</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">2024-09-28</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">Market</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">1.1</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">Market Price</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">x2</p></td><td><p align="left" color="#4C972E" class="sc-blHHSb cUlyYr">buy</p></td><td><p align="left" class="sc-blHHSb bXnAcp">Filled</p></td></tr><tr class="sc-keTIit hZSLjO"><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">BTC/USD</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">2024-09-28</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">Limit</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">1.1</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">4.11</p></td><td><p align="left" color="tertiary" class="sc-blHHSb jFjBXg">x2</p></td><td><p align="left" color="#4C972E" class="sc-blHHSb cUlyYr">buy</p></td><td><p align="left" class="sc-blHHSb bXnAcp">Filled</p></td></tr></tbody></table>"`;
